README.md
<div align="center">
<img width="640px" alt="" src="https://user-images.githubusercontent.com/47371276/154186533-0ef65901-018c-4151-a5cc-c9bc2e25fc51.png" />
</div>

# modular-ui<img src="https://user-images.githubusercontent.com/47371276/153836684-89e0f709-afd8-46d5-b2b0-41cb38614809.png" align="left" alt="" width="40"/>

![NPM](https://img.shields.io/npm/l/@cianciarusocataldo/modular-ui?label=License&style=for-the-badge) 
![npm (scoped)](https://img.shields.io/npm/v/@cianciarusocataldo/modular-ui?color=orange%20&label=Latest%20version&style=for-the-badge&logo=npm) 
![npm bundle size (scoped)](https://img.shields.io/bundlephobia/min/@cianciarusocataldo/modular-ui?label=Package%20size&style=for-the-badge&logo=npm)
![Maintenance](https://img.shields.io/maintenance/yes/2025?label=Maintained&style=for-the-badge)

* * *

## Status<img align="left" alt="" src="https://user-images.githubusercontent.com/47371276/154184535-be975554-62a7-47ff-bda4-a0dea502b3a3.png" width="15"  height="30" />

### Security

![Code Climate issues](https://img.shields.io/codeclimate/issues/CianciarusoCataldo/modular-ui?label=Code%20Climate%20issues)
![Code Climate technical debt](https://img.shields.io/codeclimate/tech-debt/CianciarusoCataldo/modular-ui?label=Code%20Climate%20tech%20debt)
![Libraries.io dependency status for latest release, scoped npm package](https://img.shields.io/librariesio/release/npm/@cianciarusocataldo/modular-ui?label=Libraries.io%20dependency%20status&logo=librariesdotio)
![Snyk Vulnerabilities for npm package](https://img.shields.io/snyk/vulnerabilities/npm/@cianciarusocataldo/modular-ui?label=Snyk%20Vulnerabilities&logo=snyk)

### Continuous Integration

[![Modular-ui CI](https://github.com/CianciarusoCataldo/modular-ui/actions/workflows/modular-ui.yml/badge.svg)](https://github.com/CianciarusoCataldo/modular-ui/actions/workflows/modular-ui.yml) 
![AppVeyor](https://img.shields.io/appveyor/build/CianciarusoCataldo/modular-ui?label=AppVeyor%20build&logo=appveyor)
![CircleCI](https://img.shields.io/circleci/build/github/CianciarusoCataldo/modular-ui?label=CircleCI&logo=circleci) 
![Cirrus CI - Specific Branch Build Status](https://img.shields.io/cirrus/github/CianciarusoCataldo/modular-ui/main?label=Cirrus%20CI&logo=cirrusci) 
![Scrutinizer build (GitHub/Bitbucket)](https://img.shields.io/scrutinizer/build/g/CianciarusoCataldo/modular-ui?label=Scrutinizer%20build&logo=scrutinizer)
![Wercker CI Run](https://img.shields.io/wercker/ci/6202992970a92108002506c3?label=Wercker%20CI&logo=oracle) 

### Code analysis

![Codacy grade](https://img.shields.io/codacy/grade/7d09cb70f1844653a8b5978a3d8f164e?label=Codacy%20code%20grade&logo=codacy)
![CodeFactor Grade](https://img.shields.io/codefactor/grade/github/cianciarusocataldo/modular-ui?label=CodeFactor%20code%20quality&logo=codefactor)
![Code Climate maintainability](https://img.shields.io/codeclimate/maintainability/CianciarusoCataldo/modular-ui?label=Code%20Climate%20maintainability&logo=codeclimate)
[![Language grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/CianciarusoCataldo/modular-ui.svg?logo=lgtm&logoWidth=18&label=LGTM%20code%20quality)](https://lgtm.com/projects/g/CianciarusoCataldo/modular-ui/context:javascript) 
![Scrutinizer code quality (GitHub/Bitbucket)](https://img.shields.io/scrutinizer/quality/g/CianciarusoCataldo/modular-ui?label=Scrutinizer%20code%20quality&logo=scrutinizer)
[![Reviewed by Hound](https://img.shields.io/badge/Reviewed_by-Hound-8E64B0.svg)](https://houndci.com)

### Tests coverage

![Codacy coverage](https://img.shields.io/codacy/coverage/7d09cb70f1844653a8b5978a3d8f164e?label=Codacy%20coverage&logo=codacy)
![Code Climate coverage](https://img.shields.io/codeclimate/coverage/CianciarusoCataldo/modular-ui?label=Code%20Climate%20coverage&logo=codeclimate)
![Codecov](https://img.shields.io/codecov/c/github/CianciarusoCataldo/modular-ui?label=CodeCov%20coverage&logo=codecov)
![Coveralls](https://img.shields.io/coveralls/github/CianciarusoCataldo/modular-ui?label=Coveralls%20coverage&&logo=coveralls) 
![Scrutinizer coverage (GitHub/BitBucket)](https://img.shields.io/scrutinizer/coverage/g/CianciarusoCataldo/modular-ui/main?label=Scrutinizer%20coverage)

* * *

<br>

## Summary<img align="left" alt="" src="https://user-images.githubusercontent.com/47371276/153879131-3e4416a4-b3fd-4208-beee-ccf672345bcc.png" width="30"/>

-   [Modular-ui philosophy](#modular-ui-philosophy)
-   [Components building process](#components-building-process)
    -   [Shared components properties](#shared-components-properties)
    -   [Customizable UI](#customizable-ui)
    -   [Box components](#box-components)
-   [Getting started](#getting-started)
    -   [Installation](#installation)
    -   [Usage](#usage)
-   [Tests](#tests)
-   [Authors](#authors)
-   [License](#license)

<br>

* * *

## Modular-ui philosophy

This library is built upon few (but important) concepts:

-   Small library size, to not impact on final app bundle size
-   Few dependencies, to not introduce a bunch of additional packages (that could also introduce bugs or version conflicts with pre-existent packages)
-   Let the user customize every part as desired, but with default values set otherwise, to make every component ready to be used without so many parameters
-   Every component should be integrable into every react app, so their behaviour must be customizable too (with callbacks, for example)

The same concepts are also the base of another project I maintain, [modular-app](https://github.com/CianciarusoCataldo/modular-app), a ready to use react app template that use this library internally (check it out, it is also the base of [Modular-ui guide page](https://cianciarusocataldo.github.io/modular-ui/) !)

<br>

## Components building process

This library use a standardized process to build every component. As result, every component has a shared initial logic, shared CSS styles and shared properties.

<br>

### Shared components properties

Some properties are shared between all components, for a smoother dev experience. In addition, this makes every single component easily re-usable. Let's see them in details:

| Property    | Description                                                                                                         |
| ----------- | ------------------------------------------------------------------------------------------------------------------- |
| `className` | custom className applied on main container                                                                          |
| `dark`      | Enable/disable dark mode                                                                                            |
| `hide`      | Hide/show component                                                                                                 |
| `id`        | `data-id` parameter (for testing purpose, to easily find the component into the DOM)                                |
| `shadow`    | Enable/disable shadow behind component                                                                              |
| `style`     | [Css inline properties](https://www.w3schools.com/html/html_css.asp) applied on main container                      |
| `unstyled`  | If `true`, no standard modular-ui styles will be applied on the components (useful for example, with image buttons) |

With these shared properties is possible to drive every component behaviour and UI with an external state management system ([modular](https://github.com/CianciarusoCataldo/modular) library uses these components internally in conjunction with [redux](https://redux.js.org/))

<br>

### Customizable UI

Modular-ui components UI is globally configurable, with CSS variables. By defining some specific custom CSS variables into your app css, you'll change the UI of all components:

| CSS variable                              | Description                                                                                                 | Default value                                 | Related CSS attribute |
| ----------------------------------------- | ----------------------------------------------------------------------------------------------------------- | --------------------------------------------- | --------------------- |
| `--modular-custom-background-dark`        | Components background, when in dark-mode ([dark](#shared-components-properties) is set to true).            | `linear-gradient(to right, #2d3748, #1d232e)` | `background-image`    |
| `--modular-custom-background-color-dark`  | Components background color, when in dark-mode ([dark](#shared-components-properties) is set to true).      | `#1d232e`                                     | `background-color`    |
| `--modular-custom-text-color-dark`        | Components text color, when in dark-mode ([dark](#shared-components-properties) is set to true).            | `white`                                       | `text-color`          |
| `--modular-custom-background-light`       | Components background, when not in dark-mode ([dark](#shared-components-properties) is set to false).       | `linear-gradient(to right, #fff, #eaebec)`    | `background-image`    |
| `--modular-custom-background-color-light` | Components background color, when not in dark-mode ([dark](#shared-components-properties) is set to false). | `#f5f5f5`                                     | `background-color`    |
| `--modular-custom-text-color-light`       | Components text color, when not in dark-mode ([dark](#shared-components-properties) is set to false).       | `#1b1b1b`                                     | `text-color`          |

If you don't set these variables, the default values will be used.

<br>

### Box components

Some components are designed with a specific structure, to better control and customize their look and feel. This type of components is called `Box Component`, for their particular structure. In addition to the [shared properties](#shared-components-properties), as they are part of modular-ui, they accept 3 additional (and optional) properties:

-   `label`, a string or a component rendered above the final component
-   `value`, a specific property that drive the component UI, its type vary depends on component itself (for example, for the CheckBox component, it is the check status, as a `boolean` ) 
-   `defaultValue`, the default value to use when `value` is not given or is `undefined`

<div align="center">
<img width="380px" alt="" src="https://user-images.githubusercontent.com/47371276/153539556-be82468f-b7c1-4753-8fe6-c832c35b7027.png" />
</div>

<br>

* * *



## Getting started<img src="https://user-images.githubusercontent.com/47371276/154059478-c044b278-5ef4-4036-8113-a18531e35194.svg" align="left" width="30" />


### Installation

If you want to use this library inside your project, just install it:

```

npm i @cianciarusocataldo/modular-ui

```

### Usage

After installation, you can use every `modular-ui` component in your app. Run this example to see them in action:

```tsx
import { Card, Container, Link } from "@cianciarusocataldo/modular-ui";
import { render } from "react-dom";

render(<Container animated>
          <Card
             dark={true}
             body={<p>This page is entirely made with modular-ui components !</p>}
             footer={<Link to="https://cianciarusocataldo.github.io/modular-ui">Modular-ui page</Link>}
          />
       </Container>,     
       document.getElementById("root"));

```

[![Edit modular-ui playground](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/modular-ui-playground-gfpbb?fontsize=14&hidenavigation=1&theme=dark&view=preview)

<br>

If you want to customize the UI globally, initialize the dedicated [CSS variables](#customizable-ui):

```css
*{
  --modular-text-color : #f5f5f5;
  --modular-background-color: #1b1b1b;
 }

```

* * *

## Tests<img align="left" alt="" src="https://user-images.githubusercontent.com/47371276/153876497-124aa299-63d9-46bf-a090-6d53a7417f4d.png" width="30"/>

Unit tests for every component are located inside `tests` folder. The test script is executed with pre-defined test command:

    npm run test

This lib is analyzed and tested with different automated code analysis systems:

-   [AppVeyor](https://ci.appveyor.com/project/CianciarusoCataldo/modular-ui)
-   [CircleCI](https://app.circleci.com/pipelines/github/CianciarusoCataldo/modular-ui)
-   [Cirrus CI](https://cirrus-ci.com/github/CianciarusoCataldo/modular-ui)
-   [Codacy](https://app.codacy.com/gh/CianciarusoCataldo/modular-ui)
-   [Code Climate](https://codeclimate.com/github/CianciarusoCataldo/modular-ui)
-   [CodeCov](https://app.codecov.io/gh/CianciarusoCataldo/modular-ui)
-   [CodeFactor](https://www.codefactor.io/repository/github/cianciarusocataldo/modular-ui)
-   [Coveralls](https://coveralls.io/github/CianciarusoCataldo/modular-ui)
-   [Hound](https://houndci.com/)
-   [Scrutinizer](https://scrutinizer-ci.com/g/CianciarusoCataldo/modular-ui/)
-   [Wercker](https://app.wercker.com/CianciarusoCataldo/modular-ui/runs)

<br>

* * *

## Authors<img align="left" src="https://user-images.githubusercontent.com/47371276/154069439-eece1dc0-a82c-4322-95f4-cf2211226fcf.png" alt="" width="30" />

-   [**Cataldo Cianciaruso**](https://github.com/CianciarusoCataldo)

<br>

* * *                       

## License<img align="left" src="https://user-images.githubusercontent.com/47371276/154152300-69b2ebbd-2461-49a7-b1c9-3570f4b51156.png" alt="" width="30" />

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details


CHANGELOG.md
# Modular-ui changelog

## Versions

- [5.1.1](#511)
- [5.1.0](#510)
- [5.0.1](#501)
- [5.0.0](#500)
- [4.2.0](#420)
- [4.1.0](#410)
- [4.0.3](#403)
- [4.0.2](#402)
- [4.0.0](#400)
- [3.1.0](#310)
- [3.0.0](#300)
- [2.0.0](#200)
- [1.2.0](#120)
- [1.1.0](#110)
- [1.0.0](#100)
- [0.9.4](#094)
- [0.9.0](#090)
- [0.8.0](#080)
- [0.7.0](#070)
- [0.6.0](#060)
- [0.5.0](#050)

<br>

---

## Changes

<br>

### 5.1.1

- Customizable states for `Spinner` component

<br>

### 5.1.0

- `Spinner` component

<br>

### 5.0.1

- `Drawer` logo prop removed

<br>

### 5.0.0

- Fully revisited `Form` component
- Refactoring
- `Drawer` component position is now driven by a new prop, `position` (`bottom`, `top`, `right`, `left`, `bottom-left`, `bottom-right`, `top-left` and `top-right`)
- `Modal` close button is now hidden when no `onClose` prop is passed

<br>

### 4.2.0

- `List` component

<br>

### 4.1.0

- Dropdown better styled
- Minor bugfixes
- Cleanup
- Refactoring

<br>

### 4.0.3

- Docs updated
- Styles fix

<br>

### 4.0.2

- Code properly formatted
- Styles fix
- Integrations with various CI and analysis system

<br>

### 4.0.0

- `Toggle` multi icon support, and type definition updates
- `Card` types cleaned
- `Drawer` types updates
- Refactoring
- All components type definitions exported, to make easier their integration into another project
- Components UI completely customizable by defining custom CSS variables into app css:
  - `--modular-custom-background-dark`
  - `--modular-custom-background-color-dark`
  - `--modular-custom-text-color-dark`
  - `--modular-custom-background-light`
  - `--modular-custom-background-color-light`
  - `--modular-custom-text-color-light`

<br>

### 3.1.0

- `Drawer` ease-out transition fix
- Clean up

<br>

### 3.0.0

- New `Container` component
- New `Label` component
- `Header` component removed
- `Page` component removed
- Clean up

<br>

### 2.0.0

- `Rater` component
- `Counter` component
- `Carousel` component
- Clean up
- Refactoring
- Shared component building process, with new common props
- Fully documented components, with working examples

<br>

### 1.2.0

- Bugfixes
- Dark mode driven by wrapper with CSS variables
- Clean up
- Styles better optimized
- Refactoring

<br>

### 1.1.0

- Bugfixes
- `Toggle` component
- Shadow feature for all components
- Clean up
- Refactoring

<br>

### 1.0.0

- Bugfixes
- `CodeBox` component highlight feature with Js, Python and Terminal highlight basic support
- Dark mode and Hide features available for every component
- Better styled components
- Test-suites system improved

<br>

### 0.9.4

- Bugfixes
- New `Input` component
- New `CheckBox` component
- New `CodeBox` component

<br>

### 0.9.0

- New `Table` component
- styles clean up

<br>

### 0.8.0

- New unit testing system
- Re-designed playground

<br>

### 0.7.0

- SVG moved inside components
- refactoring
- Minor fixes

<br>

### 0.6.0

- Clean up and optimizations

<br>

### 0.5.0

- Modular-ui library setup


scripts/update-version.js
const fs = require("fs");
const PKG = require("../package.json");

fs.writeFileSync(
  "./CHANGELOG.md",
  `# Modular-ui changelog

## Versions

-   [${PKG.version}](#${PKG.version})
-   [4.2.0](#420)
-   [4.1.0](#410)
-   [4.0.3](#403)
-   [4.0.2](#402)
-   [4.0.0](#400)
-   [3.1.0](#310)
-   [3.0.0](#300)
-   [2.0.0](#200)
-   [1.2.0](#120)
-   [1.1.0](#110)
-   [1.0.0](#100)
-   [0.9.4](#094)
-   [0.9.0](#090)
-   [0.8.0](#080)
-   [0.7.0](#070)
-   [0.6.0](#060)
-   [0.5.0](#050)

<br>

* * *

## Changes

<br>

### ${PKG.version}

-   

<br>

### 4.2.0

-   List component

<br>

<br>

### 4.1.0

-   Dropdown better styled
-   Minor bugfixes
-   Cleanup
-   Refactoring

<br>

### 4.0.3

-   Docs updated
-   Styles fix

<br>

### 4.0.2

-   Code properly formatted
-   Styles fix
-   Integrations with various CI and analysis system

<br>

### 4.0.0

-   \`Toggle\` multi icon support, and type definition updates
-   \`Card\` types cleaned
-   \`Drawer\` types updates
-   Refactoring
-   All components type definitions exported, to make easier their integration into another project
-   Components UI completely customizable by defining custom CSS variables into app css:
    -   \`--modular-custom-background-dark\`
    -   \`--modular-custom-background-color-dark\`
    -   \`--modular-custom-text-color-dark\`
    -   \`--modular-custom-background-light\`
    -   \`--modular-custom-background-color-light\`
    -   \`--modular-custom-text-color-light\`

<br>

### 3.1.0

-   \`Drawer\` ease-out transition fix
-   Clean up

<br>

### 3.0.0

-   New \`Container\` component
-   New \`Label\` component
-   \`Header\` component removed
-   \`Page\` component removed
-   Clean up

<br>

### 2.0.0

-   \`Rater\` component
-   \`Counter\` component
-   \`Carousel\` component
-   Clean up
-   Refactoring
-   Shared component building process, with new common props
-   Fully documented components, with working examples

<br>

### 1.2.0

-   Bugfixes
-   Dark mode driven by wrapper with CSS variables
-   Clean up
-   Styles better optimized
-   Refactoring

<br>

### 1.1.0

-   Bugfixes
-   \`Toggle\` component
-   Shadow feature for all components
-   Clean up
-   Refactoring

<br>

### 1.0.0

-   Bugfixes
-   \`CodeBox\` component highlight feature with Js, Python and Terminal highlight basic support
-   Dark mode and Hide features available for every component
-   Better styled components
-   Test-suites system improved

<br>

### 0.9.4

-   Bugfixes
-   New \`Input\` component
-   New \`CheckBox\` component
-   New \`CodeBox\` component

<br>

### 0.9.0

-   New \`Table\` component
-   styles clean up

<br>

### 0.8.0

-   New unit testing system
-   Re-designed playground

<br>

### 0.7.0

-   SVG moved inside components
-   refactoring
-   Minor fixes

<br>

### 0.6.0

-   Clean up and optimizations

<br>

### 0.5.0

-   Modular-ui library setup
`
);
